### PR TITLE
[CELEBORN-1256] Added internal port and auth support to Celeborn worker.

### PR DESCRIPTION
**Pull request by otterc**
### What changes were proposed in this pull request?

This adds an internal port and auth support to Celeborn Wokers.
1. Internal port is used by a worker to receive messages from Celeborn Master.
2. Authentication support for secure communication with clients. This change doesn't add the support in clients to communicate to the Workers securely. That will be in a future change.

This change targets just adding the port and auth support to Worker. The following items from the proposal are still pending:

- Persisting the app secrets in Ratis.
- Forwarding secrets to Workers and having ability for the workers to pull registration info from the Master.
- Secured communication between workers and clients.

### Why are the changes needed?
It is needed for adding authentication support to Celeborn ([CELEBORN-1011](https://issues.apache.org/jira/browse/CELEBORN-1011))

### Does this PR introduce _any_ user-facing change?
Yes

### How was this patch tested?
Part of a bigger change. For this change, only modified existing UTs.

### CODE REVIEW

**Comment by otterc** [comment]
@RexXiong @FMX @pan3793 @mridulm @akpatnam25 @rmcyang @waitinfuture
Could you please help review?

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala [hunk 1]
```diff
@@ -156,7 +218,7 @@ private[celeborn] class Worker(
         pushServerLimiter,
         conf.workerPushHeartbeatEnabled,
         workerSource)
-    transportContext.createServer(conf.workerPushPort)
+    transportContext.createServer(conf.workerPushPort, getServerBootstraps(transportConf))
```
Comment (otterc) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:221]: We will need separate securedPorts for push and fetch as well.

**Comment by otterc** [comment]
We will need separate secured ports for push and fetch for allowing both authenticated and un-authenticated clients to be able to communicate.

File: common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java [hunk 2]
```diff
@@ -59,6 +59,8 @@ public static String getFileName(String uniqueId, Mode mode) {
   private int pushPort;
   private int fetchPort;
   private int replicatePort;
+  private int internalPort;
+  private int securedPort;
```
Comment (mridulm) [inline common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java:63]: Can we add a note on when which port is to be used/is valid ?
Either here, or as javadoc to the `get*Port` methods below ?

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala [hunk 3]
```diff
@@ -86,19 +90,40 @@ private[celeborn] class Worker(
     conf,
     Math.min(64, Math.max(4, Runtime.getRuntime.availableProcessors())))
```
Comment (mridulm) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:92]: Are we going to support for disabling the existing behavior via a config in a future pr ? (`workerArgs.port` - which is not secure and is used for both s2s and c2s connections)

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala (hunk 1 above)
Comment (mridulm) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:221]: I was thinking the same while reviewing ...

**Review by mridulm — COMMENTED** [review]
Took a quick pass through it.

Given the number of ports, I am wondering if we should encapsulate all port configs in a class in itself - and localize all port related validation and config there (including RpcEnv creation, etc)

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala (hunk 3 above)
Comment (otterc) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:92]: Yes, will be making that change in a future pr. I haven't created a jira yet.

Comment (otterc) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:92]: Created https://issues.apache.org/jira/browse/CELEBORN-1268 to address this.

**Comment by otterc** [comment]
> Given the number of ports, I am wondering if we should encapsulate all port configs in a class in itself - and localize all port related validation and config there (including RpcEnv creation, etc)

Can this be done in a separate change, as it will require refactoring of existing code?

**Comment by otterc** [comment]
@RexXiong @FMX @pan3793 @mridulm @akpatnam25 @rmcyang @waitinfuture
This PR is ready for review now.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 4]
```diff
@@ -4624,4 +4632,37 @@ object CelebornConf extends Logging {
         "Allowed pattern is: `<host1>:<port1>[,<host2>:<port2>]*`")
       .createWithDefaultString(s"<localhost>:19097")
 
+  val WORKER_INTERNAL_PORT: ConfigEntry[Int] =
+    buildConf("celeborn.worker.internal.port")
+      .categories("worker")
+      .doc(
+        "Internal server port on the Worker where the master nodes connect.")
```
Comment (melonist) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4639]: (super nit) this and L4647-4748 can be consolidated to a single line.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 5]
```diff
@@ -4624,4 +4632,37 @@ object CelebornConf extends Logging {
...
+      .version("0.5.0")
+      .intConf
+      .createWithDefault(0)
```
Comment (melonist) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4642]: For these new ports, is 0 the correct default value? I noticed that other ports in this conf have rather specific default values.

File: common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala [hunk 6]
```diff
@@ -126,13 +147,19 @@ class WorkerInfo(
...
-      s"FetchPort:$fetchPort:ReplicatePort:$replicatePort"
+  def readableAddress(authEnabled: Boolean): String = {
+    if (authEnabled) {
+      s"Host:$host:RpcPort:$rpcPort:PushPort:$pushPort:" +
+        s"FetchPort:$fetchPort:ReplicatePort:$replicatePort:InternalPort:$internalPort" +
+        s":SecuredRpcPort:$securedRpcPort:SecuredPushPort:$securedPushPort:SecuredFetchPort:$securedFetchPort"
+    } else {
+      s"Host:$host:RpcPort:$rpcPort:PushPort:$pushPort:" +
+        s"FetchPort:$fetchPort:ReplicatePort:$replicatePort:InternalPort:$internalPort"
+    }
```
Comment (melonist) [inline common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:158]: This could be simplified into the following:
```
val addressStr = s"Host:$host:RpcPort:$rpcPort:PushPort:$pushPort:" +
        s"FetchPort:$fetchPort:ReplicatePort:$replicatePort:InternalPort:$internalPort"
if (authEnabled) {
    addressStr +=  s":SecuredRpcPort:$securedRpcPort:SecuredPushPort:$securedPushPort:SecuredFetchPort:$securedFetchPort"
}
addressStr
```

File: common/src/main/scala/org/apache/celeborn/common/protocol/message/ControlMessages.scala [hunk 7]
```diff
@@ -548,6 +572,8 @@ object ControlMessages extends Logging {
         .setHighWorkload(highWorkload)
         .setWorkerStatus(PbSerDeUtils.toPbWorkerStatus(workerStatus))
         .setRequestId(requestId)
+        .setInternalPort(internalPort)
+        .setSecuredRpcPort(securedRpcPort)
         .build().toByteArray
```
Comment (melonist) [inline common/src/main/scala/org/apache/celeborn/common/protocol/message/ControlMessages.scala:577]: Should this also have .setSecuredPushPort(securedPushPort) and .setSecuredFetchPort(securedFetchPort)?

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 5 above)
Comment (otterc) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4642]: For the corresponding un-secured ports, the default is 0 as well. When it is 0, then the system picks an ephemeral port. For these ports, that is desired.

File: common/src/main/scala/org/apache/celeborn/common/protocol/message/ControlMessages.scala (hunk 7 above)
Comment (otterc) [reply common/src/main/scala/org/apache/celeborn/common/protocol/message/ControlMessages.scala]: Good catch! fixed.

File: common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java [hunk 8]
```diff
@@ -281,12 +365,26 @@ public boolean equals(Object other) {
         && host.equals(o.host)
         && rpcPort == o.rpcPort
         && pushPort == o.pushPort
-        && fetchPort == o.fetchPort;
+        && fetchPort == o.fetchPort
+        && internalPort == o.internalPort
```
[the remark this answers is not in the document]
Comment (mridulm) [reply common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java:369]: Interesting idea.
QQ: currently the 'identity' is tied to host/ports - and so shipped around in worker info and/or partition location.
Is there an indirection to this ? (in spark, for example, executor id - which encapsulates a host/port for example).

If yes, we can put all the host/ports behind that - and have only the identity propagated as part of the messages ? Thoughts ?

[the remark this answers is not in the document]
Comment (otterc) [reply common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java:369]: > I think we can reuse rpcPort/pushPort/fetchPort instead of introducing secure companions

@waitinfuture So the reason I introduced these ports was because in our last discussion, we reached the conclusion that we will have different ports for secured connection than unsecured communication. Otherwise, that will be backward compatibility issues between clients that are still on older version and servers which have been upgraded.

You are right about `internalPort`. It can be removed from `PartitionLocation`. Since, I had to add other ports, I added that as well.

> Is it necessary to put internalPort/securedRpcPort/securedPushPort/securedFetchPort inside PartitionLocation? This can cause backward compatibility issue when rolling upgrade where previously auth is disabled but after upgrading auth is enabled (same to WorkerInfo). 

About this, my understanding is that we can achieve this without any issues is by first upgrading the binaries of both servers and clients without turning on authentication on either side. Once, this is done, then auth can be turned on the servers. 

I am also open to other suggestions as @mridulm pointed out.

[the remark this answers is not in the document]
Comment (waitinfuture) [reply common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java:369]: @mridulm Unfortunately there is no indirection to PartitionLocation or WorkerInfo for now. As you said, Worker is identified by host/ports, and PartitionLocation is identified by uniqueId plus WorkerInfo. This has worked well for long. We may not want to add a UUID to PartitionLocation since the message is quite big already when there are lots of PartitionLocations.

@otterc For the backward compatibility, I think you are right about PartitionLocations, but for WorkerInfos, the restarted new Workers have secured ports that are not equal to zero, so when it registers itself to Master, Master will think it's a new Worker, this may lead to issues.

> So the reason I introduced these ports was because in our last discussion
I totally agree to introduce these secure ports. What I mean is that we don't need to add new variables in PartitionLocation, instead we can reuse the existing variables (rpcPort/pushPort/fetchPort) to store the actual secured ports if auth is enabled.

[the remark this answers is not in the document]
Comment (otterc) [reply common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java:369]: > but for WorkerInfos, the restarted new Workers have secured ports that are not equal to zero, so when it registers itself to Master, Master will think it's a new Worker, this may lead to issues.

Is this because Master maintains some state in Ratis for each worker? I would think this would be a problem even now because by default we are using ephemeral ports for Workers. Every time the worker is restarted, it will use different ports for RPC, push, and fetch. Let me know if I am missing something.

> What I mean is that we don't need to add new variables in PartitionLocation, instead we can reuse the existing variables (rpcPort/pushPort/fetchPort) to store the actual secured ports if auth is enabled.

Got it. For this, the servers would then have to maintain a state whether an app is secured or not. I think this is needed anyways if we want to prevent an app that securely talks to Master but tries to communicate with the workers on the unsecured port (CELEBORN-1261). I can go ahead with this. WDYT @mridulm?

[the remark this answers is not in the document]
Comment (waitinfuture) [reply common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java:369]: > Every time the worker is restarted, it will use different ports for RPC, push, and fetch. 

This is in fact fine because workers with the same ip but different ports are considered different workers.

[the remark this answers is not in the document]
Comment (mridulm) [reply common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java:369]: There are a few combinations here we need to ensure work fine - for example, some subset of apps which are secured while others are not, etc @otterc

[the remark this answers is not in the document]
Comment (mridulm) [reply common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java:369]: @waitinfuture Interesting, after the authn work, let us relook at whether we can make the existing message sizes more compact if possible. Thanks for the details.

[the remark this answers is not in the document]
Comment (otterc) [reply common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java:369]: > There are a few combinations here we need to ensure work fine - for example, some subset of apps which are secured while others are not

I think removing the changes to `PartitionLocation` from this PR and then addressing it in a separate JIRA might be better? Sending `PartitionLocations` to client based on whether the app is secured or unsecured, maybe more complex. We can decide which approach to take in a followup change. WDYT @mridulm @waitinfuture

[the remark this answers is not in the document]
Comment (mridulm) [reply common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java:369]: Sounds good to me.

[the remark this answers is not in the document]
Comment (waitinfuture) [reply common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java:369]: > Why would this be an issue if the worker is restarted with authentication enabled? The earlier instance stops and de-registers itself, and then a new worker instance comes up with different ports. Why does the Master need to associate it with the previous instance?

This relates to graceful shutdown. If `celeborn.worker.graceful.shutdown.enabled` is enabled and workers are configured with fixed ports, when restart Worker, it will save its in-memory states(mostly PartitionLocations) in local leveldb(or rocksdb) and restore after restart, so it can serve shuffle read normally without interfering the application. This feature requires that workers are configured with fixed ports. If the Master thinks it is a new worker, after `celeborn.master.heartbeat.worker.timeout` the Master will mark the previous workers as lost and return to LifecycleManager through `HeartbeatFromApplicationResponse` and propagate to ShuffleClient, this may cause an issue.

[the remark this answers is not in the document]
Comment (waitinfuture) [reply common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java:369]: More on rolling upgrade and graceful shutdown here https://celeborn.apache.org/docs/latest/upgrading/

[the remark this answers is not in the document]
Comment (otterc) [reply common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java:369]: Ok. Then we can incorporate this logic for `WorkerInfos` equality:

```
if (celeborn.worker.graceful.shutdown.enabled  &&  !disable_unsecured_ports) 
       then WorkerInfos equality should exclude the secured ports
else if (disable_unsecured_ports) 
       then WorkerInfos equality should exclude the unsecured ports
else 
      include all the ports.
```
Note: disable_unsecured_ports config is not added yet.
cc. @waitinfuture @mridulm

[the remark this answers is not in the document]
Comment (waitinfuture) [reply common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java:369]: Sounds good to me, thanks! @otterc

File: common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala [hunk 9]
```diff
@@ -36,6 +36,10 @@ class WorkerInfo(
     val pushPort: Int,
     val fetchPort: Int,
     val replicatePort: Int,
+    val internalPort: Int,
```
Comment (RexXiong) [inline common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:39]: I don't think we need to add these secure ports(It looks a little complicated beside insecure ports), IMO the celeborn service is either open securely or not. It seems strange to keep secure connection beside with insecure connection, if so it's still insecure for celeborn service?

For users who want to use security features, they can create a new Celeborn service to instead and finally migrate other tasks to it.

Comment (waitinfuture) [reply common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:39]: @RexXiong Interesting, so you mean that we don't need to support seamless rolling upgrade from previously unsecured cluster to secured cluster, and one cluster can be either secured or unsecured, not both. Supporting mixed mode for rolling upgrade does introduces complexity. WDYT @otterc @mridulm @pan3793 ?

Comment (otterc) [reply common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:39]: In our last sync about authentication on September 23rd, we concluded, based on community feedback, that we would add separate ports for secured communication to allow for seamless rolling upgrades. In that meeting, we decided that it would be akin to supporting both HTTP and HTTPS by a web server. Here is the link to the summary: https://www.mail-archive.com/dev@celeborn.apache.org/msg00618.html.

However, IMO, allowing both secured and unsecured communication adds complexity and creates the possibility of issues. There is a way to achieve seamless rolling upgrades without adding new secured ports to Celeborn: by deploying another cluster of Celeborn that is secured (with different configured ports for Master/Worker) and leaving the existing unsecured cluster as it is. However, this creates more toil for the deployment/monitoring infrastructure.

We are not currently running Celeborn in production, and we will only run it in production with authentication enabled. Therefore, seamless rolling upgrades are not a problem for us, but we were addressing the community's concern.

If the community agrees that adding dedicated secured ports is not something we want to proceed with, I am willing to make those changes.

cc. @waitinfuture @mridulm @pan3793 @RexXiong

Comment (waitinfuture) [reply common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:39]: Thanks @otterc for the explanation. I agree with you that 

> allowing both secured and unsecured communication adds complexity and creates the possibility of issues

, and I believe @RexXiong have the same opinion. IMO whether to support rolling upgrade should depend on whether auth is enabled, as the following form describes:

| Auth enabled before  | Auth enabled after | Should support rolling upgrade |
| ------------- | ------------- | ------------- |
| No  | No  | Yes |
| No  | Yes  | No |
| Yes  | Yes  | Yes |
| Yes  | No  | No |

In summary, we should support rolling upgrade only if auth configuration stays the same before and after, and don't support otherwise.

WDYT? @otterc @RexXiong @mridulm @pan3793

Comment (RexXiong) [reply common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:39]: Thanks @waitinfuture @otterc, sorry I may have lost track of some aspects of the original design, but this really adds a lot of complexity, I agree with summary about `we should support rolling upgrade only if auth configuration stays the same`

Comment (pan3793) [reply common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:39]: it seems there is more complexity than I first thought.

given the above discussion plus feedback from Celeborn users in other threads, most community members tend to downgrade the backward compatibility to "support rolling upgrade only if auth configuration stays the same", so I'm fine with this decision

Comment (mridulm) [reply common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:39]: Thanks, as @otterc mentioned, this will significantly reduce the complexity !

Comment (otterc) [reply common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:39]: Sounds good to me. Will remove the additional secure ports. We have already committed a secure RPC port for Master for which I will create a separate PR to remove. Please note that we do need internal ports for the servers to communicate with each other. The internal ports allow internal communication to follow a different flow than the external. However, the internal port number doesn't need to be part of the WorkerInfos equality.

Comment (otterc) [reply common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:39]: I have made the necessary changes. Please take a look.

**Comment by otterc** [comment]
@waitinfuture @mridulm @pan3793 @RexXiong I have updated the PR. Please take a look.

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks! cc @RexXiong PTAL

**Review by SteNicholas — APPROVED** [review]

**Review by FMX — APPROVED** [review]

**Comment by waitinfuture** [comment]
Thanks! Merged to main(v0.5.0)
